lib/spack/spack/spec.py

```diff
@@ -4554,6 +4557,13 @@ def format_attribute(match_object: Match) -> str:
             color = None
             if "variants" in parts:
                 color = VARIANT_COLOR
+
+                # if the spec has a namespace and no name, include the namespace in its variants.
+                if not self.name and self.namespace and isinstance(current, vt.VariantMap):
+                    if not self.name and self.namespace:
+                        current = current.copy()
+                        current["namespace"] = vt.SingleValuedVariant("namespace", self.namespace)
```

Comment:
I'm a little skeptical of the `namespace` being treated as a variant for decisions of whether to print it or not, I'll have to grab the code and play with it a bit to figure out when printing an anonymous spec is even relevant though.

Reply:
I thought about this a bit more and I came up with something better.  I now allow a final `?` in format strings to make them evaluate to `""` if the value of the attribute formatted is `None`.  I added a `namespace_if_anonymous` property to specs and, finally, I allowed arbitrary identifiers as "sigils", along with spaces. So we can now append `{ namespace=namespace_if_anonymous?}` to the default format string to handle this case.  I think this is... somewhat better than what was there before, and it removes the special variant handling.

Thoughts on why we do this:
* We don't print the namespace by default -- we print `{name}`, not `{fullname}` in the default format string. 
* You *could* always print `fullname`, but:
    * We don't really show namespaces many places in the UI and I'm not sure how best to do it. Before this, format() on an anonymous spec with a namespace would just print `''`, which is not good. The error message from `spack find` was `"No package matches query:"`, which is clearly not what we want to print.
    * `fullname` for an anonymous spec with a namespace currently formats as, `builtin.None`, which I don't think we ever want to see. It's also not round-trippable.
    * You can't *just* write the namespace or the syntax is ambiguous.  e.g., if you just print `builtin`, spack's going to assume that's a package called `builtin`, not a namespace.
    * You could write `builtin.`, I suppose, and perhaps that should actually parse and be the format we print for `fullname`, but it is not very readable:
        ```yaml
         projections:
             myrepo.: "myrepo_special_path/{name}-{hash}"
             builtin.: "builtin/{name}-{hash}"
        ```
* Even if you fix `fullname` as above, I do not think we want it to be the default b/c people are very used to not showing it in the UI, and scripts using `spack find`, etc. are assuming that they'll get just `{name}`.
    * `spack find` with `fullname` as the default format is going to be really noisy.
    * I think this means that we should *only* print the namespace if the spec is anonymous.

Reply:
See the latest commit for details.

Reply:
I pulled this out into its own PR in #45487, since we are likely going to want to point to the format changes separately. In particular, I made `arch=` less of a special case.

I think the changes in #45487 will help with regularizing formats... and they'll allow me to finish #42333, which has been really hard to get right (because of all the special cases around spacing in format strings)